File: option.js
exports.commitMessage = function(info, userInfo) {
	if (userInfo.hasOwnProperty('commitMessage'))
		return replaceAll(userInfo.commitMessage, info.problemNumber, info.problemTitle);

	return 'https://www.acmicpc.net/problem/' + info.problemNumber;
}

exports.sourceTree = function(problemNumber, userInfo) {
	if (userInfo.hasOwnProperty('sourceTree')) {
		if (userInfo.sourceTree.endsWith('/')) {
			var sourceTree = userInfo.sourceTree.slice(0, userInfo.sourceTree.length - 1);
			return sourceTree.replace(/\[NO\]/gi, problemNumber);
		}
		
		return userInfo.sourceTree.replace(/\[NO\]/gi, problemNumber);
	}

	return userInfo.repo;
}

exports.dirName = function(problemNumber, userInfo) {
	if (userInfo.hasOwnProperty('dirName'))
		return (userInfo.dirName.replace(/\[NO\]/gi, problemNumber)) + '/';

	return problemNumber + '/';
}

exports.mkdir = function(problemNumber, userInfo) {
	if (userInfo.hasOwnProperty('mkdir')) {
		if (userInfo.mkdir)
			return exports.dirName(problemNumber, userInfo);
		else
			return '';
	}

	return exports.dirName(problemNumber, userInfo);
}

exports.private = function(info, userInfo) {
	if (userInfo.hasOwnProperty('private'))
		if (userInfo.private && info.private)
			return false;

	return true;
}

exports.poll = function(userInfo) {
	if (userInfo.hasOwnProperty('poll'))
		return userInfo.poll;

	return 600000;
}

exports.sourceName = function(problemNumber, language, userInfo) {
	if (userInfo.hasOwnProperty('sourceName'))
		return (userInfo.sourceName.replace(/\[NO\]/gi, problemNumber)) + getExt(language);

	return problemNumber + getExt(language);
}

function replaceNumber(value, problemNumber) {
	return value.replace(/\[NO\]/gi, problemNumber);
}

function replaceAll(value, problemNumber, problemTitle) {
	return value.replace(/\[NO\]/gi, problemNumber)).replace(/\[TITLE\]/gi, problemTitle));
}

function getExt(language) {
	switch (language) {
		case 'C': return '.c';
		case 'C++': return '.cc';
		case 'C++11': return '.cc';
		case 'C++14': return '.cc';
		case 'Java': return '.java';
		case 'C11': return '.c';
		case 'Python': return '.py';
		case 'Python3': return '.py';
		case 'PyPy': return '.py';
		case 'PyPy3': return '.py';
		case 'Ruby2.2': return '.rb';
		case 'C#4.0': return '.cs';
		case 'Text': return '.txt';
		case 'Go': return '.go';
		case 'F#': return '.fs';
		case 'Pascal': return '.pas';
		case 'Lua': return '.lua';
		case 'Perl': return '.pl';
		case 'C(Clang)': return '.c';
		case 'C++11(Clang)': return '.cc';
		case 'C++14(Clang)': return '.cc';
		case 'Fortran': return '.f95';
		case 'Scheme': return '.scm';
		case 'Ada': return '.ada';
		case 'awk': return '.awk';
		case 'OCaml': return '.ml';
		case 'Brainfuck': return '.bf';
		case 'Whitespace': return '.ws';
		case 'Tcl': return '.tcl';
		case 'Assembly(32bit)': return '.asm';
		case 'D': return '.d';
		case 'Clojure': return '.clj';
		case 'Rhino': return '.js';
		case 'Cobol': return '.cob';
		case 'SpiderMonkey': return '.js';
		case 'Pike': return '.pike';
		case 'sed': return '.sed';
		case 'Rust': return '.rs';
		case 'Intercal': return '.i';
		case 'bc': return '.bc';
		case 'VB.NET4.0': return '.vb';
		case '아희': return '.aheui';
		default: return '';
	}
}

File: option.py
DEFAULT_COMMIT_MESSAGE = 'BOJ #[NO]'
DEFAULT_DIR_NAME = '[NO]'
DEFAULT_POLL = 600
DEFAULT_SOURCE_NAME = '[NO]'


class Option:

    def __init__(self, option):
        self.option = option

    def commit_message(self, problem):
        if not 'commit_message' in self.option:
            return self.replace_msg(DEFAULT_COMMIT_MESSAGE, problem)

        return self.replace_msg(self.option['commit_message'], problem)

    def source_tree(self, problem, repo_name):
        if not 'source_tree' in self.option:
            if self.mkdir():
                return '%s/%s' % (repo_name, self.dir_name(problem))
            return '%s' % self.repo_name

        if self.option['source_tree'][-1] == '/':
            if self.mkdir():
                return '%s%s' % (self.option['source_tree'], self.dir_name(problem))
            return '%s' % self.option['source_tree'][:-1]

        if self.mkdir():
            return '%s/%s' % (self.option['source_tree'], self.dir_name(problem))
        return '%s' % self.option['source_tree']

    def dir_name(self, problem):
        if not 'dir_name' in self.option:
            return self.replace_msg(DEFAULT_DIR_NAME, problem)

        return self.replace_msg(self.option['dir_name'], problem)

    def mkdir(self):
        if not 'mkdir' in self.option:
            return True

        return self.option['mkdir']

    def private(self):
        if not 'private' in self.option:
            return False

        return self.option['private']

    def poll(self):
        if not 'poll' in self.option:
            return DEFAULT_POLL

        return self.option['poll']

    def source_name(self, problem):
        if not 'source_name' in self.option:
            return self.replace_msg(DEFAULT_SOURCE_NAME, problem)

        return self.replace_msg(self.option['source_name'], problem)
    
    def lang(self, problem):
        if not 'lang' in self.option:
            return False, None
        if problem['language'] != self.option['lang']:
            return True, False

        return True, True

    def replace_msg(self, msg, problem):
        msg = msg.replace('[NO]', problem['problem_id'])
        msg = msg.replace('[TITLE]', problem['problem_title'])
        return msg

    def get_ext(self, language):
        extensions = {
            'C': '.c',
            'C++': '.cpp',
            'C++11': '.cpp',
            'C++14': '.cpp',
            'C++17': '.cpp',
            'Java': '.java',
            'Java (OpenJDK)': '.java',
            'C11': '.c',
            'Python 2': '.py',
            'Python 3': '.py',
            'PyPy2': '.py',
            'PyPy3': '.py',
            'Ruby2.5': '.rb',
            'Kotlin': '.kt',
            'Swift': '.swift',
            'C# 6.0': '.cs',
            'Text': '.txt',
            'node.js': 'js',
            'Go': '.go',
            'F#': '.fs',
            'PHP': '.php',
            'Pascal': '.pas',
            'Lua': '.lua',
            'Perl': '.pl',
            'Objective-C': '.m',
            'Objective-C++': '.mm',
            'C (Clang)': '.c',
            'C++11 (Clang)': '.cpp',
            'C++14 (Clang)': '.cpp',
            'C++17 (Clang)': '.cpp',
            'Golfscript': '.gs',
            'Bash': '.sh',
            'Fortran': '.f95',
            'Scheme': '.scm',
            'Ada': '.ada',
            'awk': '.awk',
            'OCaml': '.ml',
            'Brainfuck': '.bf',
            'Whitespace': '.ws',
            'Tcl': '.tcl',
            'Assembly (32bit)': '.asm',
            'Assembly (32bit)': '.asm',
            'D': '.d',
            'Clojure': '.clj',
            'Rhino': '.js',
            'Cobol': '.cob',
            'SpiderMonkey': '.js',
            'Pike': '.pike',
            'sed': '.sed',
            'Rust': '.rs',
            'Boo': '.boo',
            'Intercal': '.i',
            'bc': '.bc',
            'Nemerle': '.n',
            'Cobra': '.cobra',
            'Algol 68': '.a68',
            'Befunge': '.bf',
            'Haxe': '.hx',
            'LOLCODE': '.lol',
            'VB.NET 4.0': '.vb',
            '아희': '.aheui'
        }
        
        if not language in extensions:
            return True, 'Unknown language'
        return False, extensions[language]
